### ISSUE TITLE
In entropy, try to compress string indices with window prediction

### ISSUE THREAD

**Issue by Yoric** [binast/binjs-ref#209]

**Comment by kannanvijayan** [comment]
The core window logic is really simple.  In the TypeScript codebase it is modeled here:

https://github.com/binast/binast-node/blob/master/src/string_cache.ts

The prediction model around this is to count the number of times a `lookup` yields a hit (i.e. `result >= 0`) or a miss (`result == -1`) and build a static probability table around that (which is done in the string-window analysis code here):

https://github.com/binast/binast-node/blob/master/src/analysis/string_window.ts

**Comment by Yoric** [comment]
As a baseline, with pure path prediction, here's what we get for frameworks/*.js 
(using its own dictionary):

    Primitives:
        bools: 3010 bytes, 105404 symbols (0.23 bits/symbol)
        floats: 4192 bytes, 15950 symbols (2.10 bits/symbol)
        unsigned longs: 6007 bytes, 24661 symbols (1.95 bits/symbol)
        string enums: 11254 bytes, 72602 symbols (1.24 bits/symbol)
    User strings:
        property keys: 46580 bytes, 52193 symbols (7.14 bits/symbol)
        identifier names: 89397 bytes, 166544 symbols (4.29 bits/symbol)
        string literals: 25133 bytes, 26175 symbols (7.68 bits/symbol)
    Composed:
        interface names: 90930 bytes, 522639 symbols (1.39 bits/symbol)
        list lengths: 21198 bytes, 98015 symbols (1.73 bits/symbol)

And for facebook/single/*.js (using its own dictionary):


    Primitives:
        bools: 45503 bytes, 1785741 symbols (0.20 bits/symbol)
        floats: 142393 bytes, 301045 symbols (3.78 bits/symbol)
        unsigned longs: 126035 bytes, 448663 symbols (2.25 bits/symbol)
        string enums: 101334 bytes, 717778 symbols (1.13 bits/symbol)
    User strings:
        property keys: 1071535 bytes, 1072868 symbols (7.99 bits/symbol)
        identifier names: 1404186 bytes, 2493885 symbols (4.50 bits/symbol)
        string literals: 1450802 bytes, 1116817 symbols (10.39 bits/symbol)
    Composed:
        interface names: 1450171 bytes, 9725561 symbols (1.19 bits/symbol)
        list lengths: 424780 bytes, 1985368 symbols (1.71 bits/symbol)

In other words, identifier names work pretty well already, but property keys and string literals not quite as much. I'll experiment both with your window mechanism and with the one I have implemented a few days ago.

**Comment by Yoric** [comment]
Switching to a first implementation of window gives result that roundtrip but are very much **not good**.

With a window width of 32:

Frameworks:

        property keys: 95365 bytes, 52193 symbols (14.62 bits/symbol)
        identifier names: 283764 bytes, 166544 symbols (13.63 bits/symbol)
        string literals: 46101 bytes, 26175 symbols (14.09 bits/symbol)

Facebook:

        property keys: 2411615 bytes, 1072868 symbols (17.98 bits/symbol)
        identifier names: 5069539 bytes, 2493885 symbols (16.26 bits/symbol)
        string literals: 2526703 bytes, 1116817 symbols (18.10 bits/symbol)

**Comment by kannanvijayan** [comment]
> So, did you actually try that, @kannanvijayan ?

I hadn't gotten around to implementing that.  It was the next set of things to do (along with number literals).

> I have difficulties convincing myself that switching to a more complicated encoding of the MISS case will somehow reduce the number of bits we need to represent it.

The justification is a bit deeper than that.  Here are the results for the 64-entry string window (attached at the end).  We see that the miss rate for identifiers is 6%, but for property names and string literals is ~40%.  Averaging across all string-table references, the miss rate is 20% of all string-refs

(this is across a bunch of framework, facebook, and mozilla JS code).

Looking at the string indices of these 40% strings, the distribution is of values is highly skewed.  The var-uint encoding of the misses is very biased towards low-values in its high component bits.

In the current varuint approach, the misses take an average of 12 bits to encode (plus the 1.3 bits required to encode the `MISS` character prefixing that).  This 12-bit average will be contributed to from string-indices which are either "small" (i.e. <128), or "large" (i.e. >=128) or "very large" (>=2^15).  Each "small" index will contribute 1 byte, "large" index 2 bytes, and "very large" index 3 bytes.

There are almost no "very large" indexes that show up.  However, for the "large" indices (two-byte) ones, show up a lot.  However, the high 7 bits of these indices are generally very likely to hold a low numeric value.

My implementation at that point would use the full 8 bits per component for the varuint encoding.  The new approach is simply to split up the components and predict each directly, with the expected gain being that the high component of the 2-byte varuint encoded stringrefs window-misses.. will go from needing 8 bits to encode, to needing 2 bits to encode.

So the napkin estimate I had for the benefits of doing this would be:

1. For about 20% of the string-refs (misses against the string-window)
2. For about the half of those that are encoded with 2 bytes (10% of total string-refs)
3. We can reduce the bit-cost per string-ref from about 18 bits to about 10 bits, nearly 40%.

So I expect about a 4% overall benefit to filesize (compared to string-window only).

[ALL.txt](https://github.com/binast/binjs-ref/files/2599908/ALL.txt)

**Comment by kannanvijayan** [comment]
Just to confirm, Yoric.. the string-window _is_ a move-to-front window right?

The text file ALL.txt attached in my previous comment shows the overall hits I was getting, which are significantly different from the ones you are reporting here.  There must be something very different about our approaches.

**Comment by Yoric** [comment]
> Just to confirm, Yoric.. the string-window _is_ a move-to-front window right?

Yes, move-to-front.

**Comment by Yoric** [comment]
> We can reduce the bit-cost per string-ref from about 18 bits to about 10 bits, nearly 40%.

Take a look at https://github.com/binast/binjs-ref/issues/208#issuecomment-439881818 . We're doing much better with Brotli.

**Comment by Yoric** [comment]
A little more statistics.

On frameworks:

    property keys: 97217 bytes, 52193 symbols (14.90 bits/symbol)
	65.19% Hits: 63751 bytes, 34026 symbols (14.99 bits/symbol)
	34.81% Misses: 33375 bytes, 18167 symbols (14.70 bits/symbol)
    identifier names: 304404 bytes, 166544 symbols (14.62 bits/symbol)
	94.12% Hits: 287173 bytes, 156746 symbols (14.66 bits/symbol)
	5.88% Misses: 17217 bytes, 9798 symbols (14.06 bits/symbol)
    string literals: 46387 bytes, 26175 symbols (14.18 bits/symbol)
	42.47% Hits: 19714 bytes, 11116 symbols (14.19 bits/symbol)
	57.53% Misses: 26693 bytes, 15059 symbols (14.18 bits/symbol)

These 14+ bits in case of hit mean that we're actually worse than just writing an uncompressed index in both HIT and MISS case. Not good.

On Facebook:

(still running)

**Comment by Yoric** [comment]
Oh, found the bug :)

Frameworks:

    property keys: 56274 bytes, 52193 symbols (8.63 bits/symbol)
        65.19% Hits: 27020 bytes, 34026 symbols (6.35 bits/symbol)
        34.81% Misses: 29258 bytes, 18167 symbols (12.88 bits/symbol)
    identifier names: 134958 bytes, 166544 symbols (6.48 bits/symbol)
        94.12% Hits: 119165 bytes, 156746 symbols (6.08 bits/symbol)
        5.88% Misses: 15804 bytes, 9798 symbols (12.90 bits/symbol)
    string literals: 33887 bytes, 26175 symbols (10.36 bits/symbol)
        42.47% Hits: 9582 bytes, 11116 symbols (6.90 bits/symbol)
        57.53% Misses: 24314 bytes, 15059 symbols (12.92 bits/symbol)

Facebook:

(still running)

**Comment by Yoric** [comment]
> So, if we use probability-tables to encode the components of the misses, instead of using a single global probability table, we should be able to reduce the cost of the misses from about 14-16 bits to <10 bits per symbol.

I am still very skeptical about this. In terms of probabilities, that would assume that a bunch of mostly random numbers somehow have higher probabilities to show up because of arbitrarily chosen digits.

> This would come at the expense of maintaining 7 static predictor tables for each string-index (one for varuint-length, and one for each component of 1-byte, 2-byte, and 3-byte varuints, respectively).

Well, since we would encode the length separately, it's actually not really varuints. Still, this would be:
- one table with 5 values for the number of digits (including a value for `null`);
- one table for 8-bit integers;
- two tables for 16-bit integers;
- three tables for 24-bit integers; (probably always empty)
- four tables for 32-bit integers. (probably always empty)

Multiplied by the number of categories that need extensibility:

- string literals;
- identifiers;
- property keys;
- floats;
- unsigned longs;
- "offsets" (which should really find a better name).

So, that's actually a total of 66 tables, with admittedly many of them empty.

> Just to clarify, are you using 3 global predictors for the string-window index hits, or contextualizing them (i.e. a string-window per path-suffix)?

No path suffix for strings in this experiment.

Although, as you can see by comparing [the baseline](https://github.com/binast/binjs-ref/issues/209#issuecomment-439200006) and [the experiment result](https://github.com/binast/binjs-ref/issues/209#issuecomment-440373163), the baseline (pure path prediction, no windows) gives much better results for the moment.
